man/man1/getcountry.1
.TH getcountry 1  "USER COMMANDS"
.SH NAME
getcountry \- extract grid cells for specified countries
.SH SYNOPSIS
.B getcountry
[\-list] [\-json]
.I countryfile gridfile countrycode ...
.SH DESCRIPTION
Program extracts country grid cells from grid file.
.SH OPTIONS
.TP
\-list
Print list of country codes and names defined.
.TP
\-json
An additional JSON metafile with suffix \fI.json\fP is written for gridfile.
.TP
.I countryfile
Country code file. Must be a JSON metafile with country map and grid file specified.
.TP
.I gridfile
Filename of grid data file written.
.TP
.I countrycode
ISO 3166-1 alpha-3 abbreviations for countries. or country codes as numbers defined in
.I include/managepar.h
.SH EXAMPLES
.TP
Create grid data file for USA:
.B getcountry
cow_full_2018.clm.json grid_usa.clm USA
.PP
.SH EXIT STATUS
.B getcountry
returns a zero exit status if the grid file is created.
Non zero is returned in case of failure.

.SH AUTHORS

For authors and contributors see AUTHORS file

.SH COPYRIGHT

(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file

.SH SEE ALSO
regridlpj(1), lpjml(1), txt2clm(1), cru2clm(1), clm(5)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project follows a [GNU standard version numbering](https://www.gnu.org/prep/standards/html_node/Releases.html#index-version-numbers_002c-for-releases)
of `major.minor.patch` with
- `major` for substantial changes to code functionality,
- `minor` for additions of modules and functionality, and
- `patch` for technical changes and bugfixes without extending the functionality of the code.

- if applicable, categories to include in the CHANGELOG.md are:
  - Added
  - Changed
  - Deprecated
  - Removed
  - Fixed
  - Security


## [Unreleased]


## [6.1.4] - 2026-07-03

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Fabian Stenzel (stenzel@pik-potsdam.de), David Hötten (davidho@pik-potsdam.de)

### Added

- Utility `cmprestart` added to compare two restart files and check for identity.
- Function `bstruct_skiparray()` and corresponding test `test_skiparray.c` added in `bstruct` library to skip arrary elements.
- Function `bstruct_gettype()` added to `bstruct` library to get datatype of object in restart file.
- Function `bstrucht_printnamestack()` added for more comprehensive error messages.
- Comparison of restart files with `cmprestart` added to merge request template.

### Changed

- Index array object is always skipped in `bstruct_findobject.c`, test `test_indexarray.c` updated to test for reading array without seeking to an index array.
- Check for seeking to position within file added to `bstruct_findobject.c`, `bstruct_open.c`, `bstruct_wopen.c` and `bstruct_seekindexarray.c`.
- Constant `MAXLEVEL` renamed to `BSTRUCT_MAXLEVEL`.
- Macros `getname()` and `isinvalidtoken()` renamed to `bstruct_getname()` and `bstruct_isinvalidtoken()`.
- String comparison in `bstruct_fprintdata` replaced by call to `isnan()` and `isinf()`.
- New token `bstruct_dzero` defined for restart files to distinguish between float and double zeros, version of bstruct files changed to 2.


## [6.1.3] - 2026-07-03

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de), Stephen Wirth(wirth@pik-potsdam.de)
- code review: Sebastian Ostberg (ostberg@pik-potsdam.de), Jannes Breier (breier@pik-potsdam.de)

### Added

- Option `-metafile` and `-json` added to `cutclm` and `catclm`.
- Option `-double` added to `cdf2soil.c` to write grid file with double datatype.

### Changed

- New datatype `Metadata` defined in `types.h` combining maps, attributes, basetemp and hlimit arrays used by `opendata()`.
- basetemp, countrymap and hlimit arrays are copied from JSON metafiles to newly created metafiles in `regridclm`, `cutlcm`, `catclm`, and `mathclm`.
- Scaling to sum of stand fractions removed for output `"vegc_avg"`.
- Utility `getcountry` reads now JSON metafile for country codes including the alpha-3 country codes and global grid instead of CLM file.
- If `-pedantic` option is set, non-matching titles in input files stops `lpjml` with an error.
- Datatype of `lon`, `lat`, and `area` elements of `Coord` changed from `Real` to `double`.
- Missing path to grid file added in `openmetafile.c`.

### Fixed

- Warning for float coordinates corrected in `cdf2coord.c`, `cdf2grid.c` and `cdf2soil.c`.
- Variables initialized to zero in `reservoir2cdf.c` and `restart2yaml.c` to avoid warnings by `gcc`.
- Carbon balance fixed in `reclaim_land.c`.
- Check for NULL pointer added in `getuser.c` to avoid SEGV if user name cannot be resolved.
- Verbose output corrected for input without time axis in `cdf2clm.c`.
- Output is rescaled by the actual sum of stand fractions in `update_daily_cell.c` and `fwriteoutput.c` to avoid division by zero (issue #411).
- Default timestep set to 1 in `regridclm.c`.
- Variable `depth` initialized to zero before layer loop in `fwriteoutput.c` to correct `soilc_1m` output.
- Length of buffer for hostname corrected in `gethost.c`.
- Output of unprintable characters corrected in `fputprintable.c`.
- Memory leak fixed in `openmetafile.c` in case of error.


## [6.1.2] - 2026-06-16

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Sibyll Schaphoff (sibylls@pik-potsdam.de), Sebastian Ostberg (ostberg@pik-potsdam.de)

### Fixed

- Check for last year for lightning and human ignition disabled in `filesexist.c`, `lpjcheck` does not return an error for standard input.


## [6.1.1] - 2026-06-12

### Contributors

author: Jannes Breier (breier@pik-potsdam.de),  Werner von Bloh (bloh@pik-potsdam.de)
code review: Christoph Müller (cmueller@pik-potsdam.de), Stephen Wirth (wirth@pik-potsdam.de)

### Added

- `logo.png` added to `man` directory
- `LICENSES` folder
- `LICENSES/CC0-1.0.md` added for artwork license

### Changed

- `AUTHORS` renamed to `AUTHORS.md` and restructured for active and previous authors
- `INSTALL` renamed to `INSTALL.md` and adjusted
- `LICENSE` renamed to `LICENSES/AGPL-3.0.md` and moved to `LICENSES` folder
- `README` renamed to `README.md` and expanded
- `configure.sh`, `src/lpj/copyright.c` and `src/lpj/printlicense.c` adjusted to reflect renaming of files
- Updated module version for `oneAPI` in `run_simulate_default.sh`


## [6.1.0] - 2026-06-12

### Contributors

- author: Luke Oberhagemann (lukeober@pik-potsdam.de), Markus Drüke (Markus.Drueke@dwd.de), Maik Billing (billing@pik-potsdam.de), Werner von Bloh (bloh@pik-potsdam.de)
- code review: Stephen Wirth (wirth@pik-potsdam.de), Christoph Müller (cmueller@pik-potsdam.de)

### Added

- Option `-inpath` added to `configure.sh` to set path for input files.
- Function `fprintclimbuf()` added in `climbuf.c` to print variables and is called in `fprintcell.c`.
- Function `fprintbuffer()` added in `buffer.c` to print variables and is called in `fprintclimbuf.c`.
- Lakes setting stored in restart file and setting compared to actual setting in `openrestart.c`.

src/utils/getcountry.c
/**************************************************************************************/
/**                                                                                \n**/
/**                  g  e  t  c  o  u  n  t  r  y  .  c                            \n**/
/**                                                                                \n**/
/**     Extract country from grid file                                             \n**/
/**                                                                                \n**/
/** (C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file    \n**/
/** authors, and contributors see AUTHORS file                                     \n**/
/** This file is part of LPJmL and licensed under GNU AGPL Version 3               \n**/
/** or later. See LICENSE file or go to http://www.gnu.org/licenses/               \n**/
/** Contact: https://github.com/PIK-LPJmL/LPJmL                                    \n**/
/**                                                                                \n**/
/**************************************************************************************/

#include "lpj.h"

#define USAGE "Usage: %s [-list] [-json] countryfile outfile country ...\n"

static int findcountryname(const char *name,
                           const Countryname countryname[],
                           int ncountries)
{
  int i;
  for(i=0;i<ncountries;i++)
    if(!strcmp(name,countryname[i].alpha_3))
      return i;
  return NOT_FOUND;
} /* 'findcountryname' */

static Bool findcountry(const int country[],int n,int c)
{
  int i;
  for(i=0;i<n;i++)
    if(country[i]==c)
      return TRUE;
  return FALSE;
} /* of 'findcountry' */

int main(int argc,char **argv)
{
  FILE *file,*out,*gridfile;
  int i,*country,n,iarg,ngrid;
  char *endptr;
  char *out_json,*arglist;
  size_t offset,grid_offset;
  Header header,outheader;
  Filename grid_name;
  Coord *grid,res;
  Coordfile coordfile;
  Type grid_type;
  Metadata metadata;
  int code;
  int format;
  float cellsize_lon,cellsize_lat;
  Bool swap,rc,isregion,isjson,islist,gridswap,ismeta;
  outheader.nyear=1;
  outheader.nstep=1;
  outheader.timestep=1;
  outheader.firstcell=0;
  outheader.order=0;
  outheader.ncell=0;
  outheader.nbands=2;
  isjson=islist=FALSE;
  for(iarg=1;iarg<argc;iarg++)
    if(argv[iarg][0]=='-')
    {
      if(!strcmp(argv[iarg],"-json"))
        isjson=TRUE;
      else if(!strcmp(argv[iarg],"-list"))
        islist=TRUE;
      else
      {
        fprintf(stderr,"Invalid option '%s'.\n"
                USAGE,argv[iarg],argv[0]);
        return EXIT_FAILURE;
      }
    }
    else
      break;
  if((islist && argc<iarg+1) || (!islist && argc<iarg+3))
  {
    fprintf(stderr,"Argument(s) missing.\n"
            USAGE,argv[0]);
    return EXIT_FAILURE;
  }
  header.datatype=LPJ_SHORT;
  header.timestep=1;
  header.nbands=1;
  header.nstep=1;
  header.order=CELLYEAR;
  header.firstcell=0;
  header.firstyear=1901;
  header.cellsize_lon=header.cellsize_lat=0.5;
  header.ncell=1;
  header.nyear=1;
  grid_type=LPJ_SHORT;
  grid_name.name=NULL;
  initmetadata(&metadata,NULL);
  file=openmetafile(&header,&metadata,&grid_name,&grid_type,&format,&swap,&offset,argv[iarg],TRUE);
  if(file==NULL)
    return EXIT_FAILURE;
  if(metadata.countrymap==NULL)
  {
    fprintf(stderr,"Country map is missing in '%s.\n",argv[iarg]);
    fclose(file);
    free(grid_name.name);
    freemetadata(&metadata);
    return EXIT_FAILURE;
  }
  if(islist)
  {
    fclose(file);
    puts("List of country codes:\nCode Name");
    for(i=0;i<metadata.countrymap_size;i++)
       printf("%s  %s\n",metadata.countrymap[i].alpha_3,metadata.countrymap[i].name);
    free(grid_name.name);
    freemetadata(&metadata);
    return EXIT_SUCCESS;
  }
  fseek(file,offset,SEEK_SET);
  if(header.nbands==1)
    isregion=FALSE;
  else if(header.nbands!=2)
  {
    fprintf(stderr,"Invalid number of bands=%d in `%s', must be 1 or 2.\n",
            header.nbands,argv[iarg]);
    fclose(file);
    freemetadata(&metadata);
    return EXIT_FAILURE;
  }
  else
    isregion=TRUE;
  if(grid_name.name==NULL)
  {
    fprintf(stderr,"Filename of grid file missing in '%s'.\n",argv[iarg]);
    fclose(file);
    freemetadata(&metadata);
    return EXIT_FAILURE;
  }
  n=argc-iarg-2;
  country=newvec(int,n);
  if(country==NULL)
  {
    printallocerr("country");
    free(grid_name.name);
    fclose(file);
    freemetadata(&metadata);
    return EXIT_FAILURE;
  }
  for(i=0;i<n;i++)
  {
    country[i]=strtol(argv[iarg+2+i],&endptr,10);
    if(*endptr!='\0')
    {
      /* argument is not a number */
      country[i]=findcountryname(argv[iarg+2+i],metadata.countrymap,metadata.countrymap_size);
      if(country[i]==NOT_FOUND)
      {
        fprintf(stderr,"Invalid number/name '%s' for country.\n",argv[iarg+2+i]);
        free(grid_name.name);
        free(country);
        fclose(file);
        freemetadata(&metadata);
        return EXIT_FAILURE;
      }
    }
  }
  freemetadata(&metadata);
  coordfile=opencoord(&grid_name,TRUE);
  if(coordfile==NULL)
  {
    free(grid_name.name);
    free(country);
    fclose(file);
    return EXIT_FAILURE;
  }
  ngrid=numcoord(coordfile);
  if(ngrid!=header.ncell)
  {
    fprintf(stderr,"Number of cells %d in grid file '%s' is different from number of cells %d in '%s'.\n",
            ngrid,grid_name.name,header.ncell,argv[iarg]);
    free(grid_name.name);
    closecoord(coordfile);
    free(country);
    fclose(file);
    return EXIT_FAILURE;
  }
  getcellsizecoord(&cellsize_lon,&cellsize_lat,coordfile);
  res.lon=cellsize_lon;
  res.lat=cellsize_lat;
  grid=newvec(Coord,numcoord(coordfile));
  if(grid==NULL)
  {
    printallocerr("grid");
    free(grid_name.name);
    closecoord(coordfile);
    free(country);
    fclose(file);
    return EXIT_FAILURE;
  }
  for(i=0;i<numcoord(coordfile);i++)
    readcoord(coordfile,grid+i,&res);
  outheader.nstep=1;
  outheader.nbands=2;
  outheader.timestep=1;
  outheader.order=CELLYEAR;
  outheader.firstyear=header.firstyear;
  outheader.datatype=grid_type;
  outheader.cellsize_lon=cellsize_lon;
  outheader.cellsize_lat=cellsize_lat;
  outheader.scalar=getcoordscale(coordfile);
  closecoord(coordfile);
  if(isjson && grid_name.fmt==META)
  {
    ismeta=TRUE;
    initmetadata(&metadata,NULL);
    gridfile=openmetafile(&outheader,&metadata,NULL,NULL,NULL,&gridswap,&grid_offset,grid_name.name,TRUE);
    if(gridfile==NULL)
      ismeta=FALSE;
    else
      fclose(gridfile);
  }
  else
    ismeta=FALSE;
  free(grid_name.name);
  outheader.firstcell=0;
  outheader.ncell=0;
  out=fopen(argv[iarg+1],"wb");
  if(out==NULL)
  {
    fprintf(stderr,"Error creating '%s': %s.\n",argv[iarg+1],strerror(errno));
    if(ismeta)
      freemetadata(&metadata);
    free(grid);
    free(country);
    fclose(file);
    return EXIT_FAILURE;
  }
  fwriteheader(out,&outheader,LPJGRID_HEADER,LPJGRID_VERSION);
  for(i=0;i<header.ncell;i++)
  {
    if(readcountrycode(file,&code,header.datatype,isregion,swap))
    {
      fprintf(stderr,"Error reading country code at %d.\n",i+1);
      if(ismeta)
        freemetadata(&metadata);
      free(grid);
      free(country);
      fclose(file);
      fclose(out);
      return EXIT_FAILURE;
    }
    if(findcountry(country,n,code))
    {
      rc=writecoord(out,grid+i,outheader.scalar,outheader.datatype);
      if(rc)
      {
        fprintf(stderr,"Error writing coordinate at %d.\n",i+1);
        if(ismeta)
          freemetadata(&metadata);
        free(grid);
        free(country);
        fclose(file);
        fclose(out);
        return EXIT_FAILURE;
      }
      outheader.ncell++;
    }
  }
  free(grid);
  free(country);
  fclose(file);
  rewind(out);
  fwriteheader(out,&outheader,LPJGRID_HEADER,LPJGRID_VERSION);
  fclose(out);
  if(isjson)
  {
    out_json=malloc(strlen(argv[iarg+1])+strlen(JSON_SUFFIX)+1);
    if(out_json==NULL)
    {
      printallocerr("filename");
      if(ismeta)
        freemetadata(&metadata);
      return EXIT_FAILURE;
    }
    strcat(strcpy(out_json,argv[iarg+1]),JSON_SUFFIX);
    arglist=catstrvec(argv,argc);
    if(arglist==NULL)
    {
      printallocerr("arglist");
      free(out_json);
      if(ismeta)
        freemetadata(&metadata);
      return EXIT_FAILURE;
    }
    out=fopen(out_json,"w");
    if(out==NULL)
    {
      printfcreateerr(out_json);
      free(arglist);
      free(out_json);
      if(ismeta)
        freemetadata(&metadata);
      return EXIT_FAILURE;
    }
    if(!ismeta)
    {
      initmetadata(&metadata,NULL);
      metadata.source=argv[0];
      metadata.variable="grid";
      metadata.unit="degree";
      metadata.long_name="cell coordinates";
    }
    fprintjson(out,argv[iarg+1],NULL,arglist,&outheader,&metadata,NULL,LPJ_SHORT,CLM,LPJGRID_HEADER,FALSE,LPJGRID_VERSION);
    if(ismeta)
      freemetadata(&metadata);
    fclose(out);
    free(out_json);
    free(arglist);
  }
  if(header.ncell)
    printf("Number of cells: %d\n",outheader.ncell);
  else
    fputs("Warning: no cells found.\n",stderr);
  return EXIT_SUCCESS;
} /* of 'main' */
